text/0000-rewards.md

```diff
@@ -0,0 +1,250 @@
+```
+/// Our subjective record of out availability transfers for this candidate.
+CandidateRewards {
+    /// Anyone who backed this parablock
+    backers: [AuthorityId; NumBackers],
+    /// Anyone who sent us chunks for this candidate
+    downloaded_from: HashMap<AuthorityId,u16>,    
+    /// Anyone to whome we sent chunks for this candidate
+    uploaded_to: HashMap<AuthorityId,u16>,
+}
+```
+We no longer require this data during disputes.  
+<!-- You could optionally track a `downloaded_one: Option<AuthorityBitField>` too, for the nodes from whome we douwnloaded only one chunk, but this seems like premature optimization -->
+
+After we approve a relay chain block, then we collect all its `CandidateRewards` into an `ApprovalsTally`, with one `ApprovalTallyRecord` for each validator.  In this, we compute `approval_usages` from the final run of the approvals loop, plus `0.8` for each backer.
+```
+/// Our subjective record of what we used from, and provided to, all other validators on the finalized chain
+pub struct ApprovalsTally(Vec<ApprovalTallyLine>);
+
+/// Our subjective record of what we used from, and provided to, all one other validators on the finalized chain
+pub struct ApprovalTallyLine {
+    /// Approvals by this validator which our approvals gadget used in marking candidates approved.
+    approval_usages: u32,
+    /// Availability chunks we downloaded from this validator for our approval checks we used.
+    used_downloads: u32,
+    /// Availability chunks we uploaded to this validator which whose approval checks we used.
+    used_uploads: u32,
+}
+```
+At finality we sum these `ApprovalsTally` for one for the whole epoch so far, into another `ApprovalsTally`.  We can optionally sum them earlier at chain heads, but this requires mutablity.
+
+### Messages
+
+After the epoch is finalized, we share the first two lines of its `ApprovalTally`.
+```
+/// Our subjective record of what we used from some other validator on the finalized chain
+pub struct ApprovalTallyMessageLine {
+    /// Approvals by this validator which our approvals gadget used in marking candidates approved.
+    approval_usages: u32,
+    /// Availability chunks we downloaded from this validator for our approval checks we used.
+    used_downloads: u32,
+}
+
+/// Our subjective record of what we used from all other validators on the finalized chain
+pub struct ApprovalsTallyMessage(Vec<ApprovalTallyMessageLine>);
+```
+
+### Rewards compoutation
+
+We compute the approvals rewards by taking the median of the `approval_usages` fields for each validator across all validators `ApprovalsTallyMessage`s.
+```
+let mut approval_usages_medians = Vec::new(); 
+for i in 0..num_validators {
+    let mut v: Vec<u32> = approvals_tally_messages.iter().map(|atm| atm.0[i].approval_usages);
+    v.sort();
+    approval_usages_medians.push(v[num_validators/2]);
+}
+```
+Assuming more than 50% honersty, these median tell us how many approval votes form each validator
```

Comment:
Yes, you can completely corrupt the rewards if you can break the 50% honesty threshold, but you can probzbly do worse things then.

I designed the elves approval gadget to be runnable on-chain, so that's an option too, even running it on a parachain, but this creates a different problem, in that people would censor how data reaches the on-chain gadget.  It seems to me this chain censorship attack is much worse in practice, because cains are pretty easy to censor.